src/frontend/src/statement/copy_table_from.rs

```diff
@@ -322,33 +361,78 @@ fn ensure_schema_matches_ignore_timezone(left: &SchemaRef, right: &SchemaRef) ->
+fn data_type_equals_ignore_timezone_with_options(
+    l: &DataType,
+    r: &DataType,
+    ts_cast: bool,
+) -> bool {
     match (l, r) {
         (DataType::List(a), DataType::List(b))
         | (DataType::LargeList(a), DataType::LargeList(b)) => {
             a.is_nullable() == b.is_nullable()
-                && data_type_equals_ignore_timezone(a.data_type(), b.data_type())
+                && data_type_equals_ignore_timezone_with_options(
+                    a.data_type(),
+                    b.data_type(),
+                    ts_cast,
+                )
         }
         (DataType::FixedSizeList(a, a_size), DataType::FixedSizeList(b, b_size)) => {
             a_size == b_size
                 && a.is_nullable() == b.is_nullable()
-                && data_type_equals_ignore_timezone(a.data_type(), b.data_type())
+                && data_type_equals_ignore_timezone_with_options(
+                    a.data_type(),
+                    b.data_type(),
+                    ts_cast,
+                )
         }
         (DataType::Struct(a), DataType::Struct(b)) => {
             a.len() == b.len()
                 && a.iter().zip(b).all(|(a, b)| {
                     a.is_nullable() == b.is_nullable()
-                        && data_type_equals_ignore_timezone(a.data_type(), b.data_type())
+                        && data_type_equals_ignore_timezone_with_options(
+                            a.data_type(),
+                            b.data_type(),
+                            ts_cast,
+                        )
                 })
         }
         (DataType::Map(a_field, a_is_sorted), DataType::Map(b_field, b_is_sorted)) => {
             a_field == b_field && a_is_sorted == b_is_sorted
         }
-        (DataType::Timestamp(l_unit, _), DataType::Timestamp(r_unit, _)) => l_unit == r_unit,
+        (DataType::Timestamp(l_unit, _), DataType::Timestamp(r_unit, _)) => {
+            l_unit == r_unit || ts_cast
+        }
+        (&DataType::Utf8, DataType::Timestamp(_, _))
+        | (DataType::Timestamp(_, _), &DataType::Utf8) => ts_cast,
         _ => l == r,
     }
 }
 
+/// Allows the file schema is a subset of table
+fn generated_schema_projection_and_compatible_file_schema(
+    file: &SchemaRef,
+    table: &SchemaRef,
+) -> (Vec<usize>, Vec<usize>, Schema) {
+    let mut file_projection = Vec::with_capacity(file.fields.len());
+    let mut table_projection = Vec::with_capacity(file.fields.len());
+    let mut compatible_fields = file.fields.iter().cloned().collect::<Vec<_>>();
+    for (file_idx, file_field) in file.fields.iter().enumerate() {
+        if let Some((table_idx, table_field)) = table.fields.find(file_field.name()) {
+            file_projection.push(file_idx);
+            table_projection.push(table_idx);
+
+            // Safety: the compatible_fields has same length as file schema
+            *compatible_fields.get_mut(file_idx).unwrap() = table_field.clone();
```

Comment:
compatible_fields is Vec, could we just access it by `[]`?
```suggestion
            compatible_fields[file_idx]= table_field.clone();
```